crates/bevy_ecs/src/query/state.rs

```diff
@@ -679,10 +724,13 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
     /// # Safety
     /// `archetype` must be from the `World` this state was initialized from.
     pub unsafe fn update_archetype_component_access(
-        &mut self,
+        &self,
         archetype: &Archetype,
         access: &mut Access<ArchetypeComponentId>,
     ) {
+        if !self.matched_archetypes.contains(archetype.id().index()) {
```

Comment:
This probably warrants a small comment: 'check if the archetype access has already been updated by an observer and if so don't bother'

Reply:
This can be removed if we land https://github.com/bevyengine/bevy/pull/16885, or at least move this logic to system param instead.